core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -647,12 +652,20 @@ class GroupCoordinator(val brokerId: Int,
     if (group.is(AwaitingSync))
       resetAndPropagateAssignmentError(group, Errors.REBALANCE_IN_PROGRESS)
 
-    group.transitionTo(PreparingRebalance)
-    info("Preparing to restabilize group %s with old generation %s".format(group.groupId, group.generationId))
-
-    val rebalanceTimeout = group.rebalanceTimeoutMs
-    val delayedRebalance = new DelayedJoin(this, group, rebalanceTimeout)
     val groupKey = GroupKey(group.groupId)
+
+    val (delay, timeout) = if (group.currentState != InitialRebalance) {
+      group.transitionTo(PreparingRebalance)
+      (group.rebalanceTimeoutMs, time.milliseconds() + group.rebalanceTimeoutMs)
+    } else {
+      joinPurgatory.cancelForKey(groupKey) match {
```

Comment:
@dguy Let me clarify a bit on my previous comment:

I felt on issue with the `prepareRebalance` and the reason that we need to introduce `initialRebalance` is that, in purgatory we only have the `tryCompleteElseWatch` function; I.e. if we call this function upon receiving the first join group request for an unknown group, the `tryComplete` part will always succeed as we only have one member now and that member has already joined. Instead by introducing the `watchForOperation` api into purgatory which does not `tryComplete` but immediately add the delayed operation into the watch list. In addition, by checking the group generation we can know if this group is new (note that only the new group has generation number 0, while a group that has completed the first join phase will then have generation number 1):

We can then do the following:

1. If a join group request is received and its group is unknown (so a new group with one member has just been created), we call `watchForOperation` in the join purgatory such that we will not check if it is completable (in fact it is) and the `tryCompleteJoin` will not be called;
2. Otherwise we call `tryCompleteElseWatch` on the join purgatory.

Then when we receive a join group request with a known group, we call `tryComplete` on the purgatory. If this is a new group, the current check in `tryCompleteJoin` will always pass since `we have n members now and all n members have joined`. Hence in addition to that check, we also need to check if `generation number` is 0: if yes, set the `new-member` field to true; otherwise do nothing.

Then in `onCompleteJoin`, which will always be triggered after the `tryCompleteJoin` if it is the new group, check if `new-member` is set; if yes, re-watch the operation by calling `watchForOperation`, otherwise proceed with the join process.

As for your question, that group goes from `Stable` -> `PreparingRebalance`, its generation number will always be larger than 1. So when we receive a join group request we won't add the delay. Admittedly this logic does not sound too simple but I felt it may generates less code.